uniffi_bindgen/src/bindings/kotlin/templates/RustBufferHelpers.kt

```diff
@@ -114,199 +261,133 @@ fun String.Companion.lift(rbuf: RustBuffer.ByValue): String {
 }
 
 fun String.Companion.read(buf: ByteBuffer): String {
-    val len = Int.read(buf)
+    val len = buf.getInt()
     val byteArr = ByteArray(len)
     buf.get(byteArr)
     return byteArr.toString(Charsets.UTF_8)
 }
 
-
-// I can't figure out how to make a generic implementation of (Any?).read, and IIUC there are some
-// restrictions on generics in Kotlin (inherited from the JVM) that make it impossible to write in the
-// style I want here. So, we use a standalone helper.
-
-fun<T> liftOptional(rbuf: RustBuffer.ByValue, readItem: (ByteBuffer) -> T): T? {
-    return liftFromRustBuffer(rbuf) { buf -> readOptional(buf, readItem) }
+fun String.lower(): RustBuffer.ByValue {
+    val byteArr = this.toByteArray(Charsets.UTF_8)
+    // Ideally we'd pass these bytes to `ffi_bytebuffer_from_bytes`, but doing so would require us
+    // to copy them into a JNA `Memory`. So we might as well directly copy then into a `RustBuffer`.
+    val rbuf = RustBuffer.alloc(byteArr.size)
+    rbuf.asByteBuffer()!!.put(byteArr)
+    return rbuf
 }
 
-fun<T> readOptional(buf: ByteBuffer, readItem: (ByteBuffer) -> T): T? {
-    if (! Boolean.read(buf)) {
-        return null
-    }
-    return readItem(buf)
+fun String.write(buf: RustBufferBuilder) {
+    val byteArr = this.toByteArray(Charsets.UTF_8)
+    buf.putInt(byteArr.size)
+    buf.put(byteArr)
 }
 
-fun<T> liftSequence(rbuf: RustBuffer.ByValue, readItem: (ByteBuffer) -> T): List<T> {
-    return liftFromRustBuffer(rbuf) { buf -> readSequence(buf, readItem) }
-}
+{% when Type::Optional with (inner_type) -%}
+{% let inner_type_name = inner_type|type_kt %}
 
-fun<T> readSequence(buf: ByteBuffer, readItem: (ByteBuffer) -> T): List<T> {
-    val len = Int.read(buf)
-    return List<T>(len) {
-        readItem(buf)
+fun lift{{ type_name }}(rbuf: RustBuffer.ByValue): {{ inner_type_name }}? {
```

Comment:
So this may be slightly controversial, and I welcome pushback here if you don't like it.

I've opted to "monomorphize" the helper functions for `Optional`, `Sequence` and `Map` types, emitting a separate helper for each concrete instance of such a type rather than a single generic helper that takes a lambda. I find this form more readable as I don't have to mentally untangle to lambda, and it happens to fit nicely with the "iterate over all the types" approach.

The downside may be slightly increased code size, although I have a hard time imagining it'll make a big difference in practice. I also think reasonable people could disagree with me about which is the more readable approach :-)

Anyway, if y'all hate this, an alternative would be to:

* Add methods on `TypeUniverse` like `has_map_types()`, `has_sequence_types()` etc.
* Change the template here to conditionally emit the Map/Sequence/Optional helpers only when those types are actually used in the interface.